### ISSUE TITLE
De-wx'ing

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#190]
All wx code must eventually end up in ~~balt~~ **gui**. This does not simply mean that one must create wrappers for wx classes/constants and drop them into gui - it means that gui must export an **API** - so gui should not return wx items on which wx methods will be called - must wrap those methods into class methods it exports  - in other words we must not only eliminate wx imports but also _wx calls on wx objects returned by gui functions_ (such as Bind). This (apart its obvious, I hope, value vis a vis code quality) will also make it easier to move to later versions of wxPython / offer an entirely different backend. There is already a great deal of it done (edited out of this issue). But there is also an awful lot left to be done. So:
- sizers - that's the biggest one. According to what I wrote above, defining say `wxEXPAND = wx.EXPAND` in gui is **not** the way to go. Rather the usage patterns of sizers in Bash must be explored from someone who groks the wx sizers API and an API be crafted in gui which defines _what_ one wants (as opposed to how to achieve it).
  - [x] `wx.LEFT etc, 0` does make any cense ?
  - [x] I would like to go through the uses of sizers a bit and see what is left out - the wx.TOP etc stuff for instance - I have a plan for adding them in the LayoutOptions
  - [x] rename default_border etc to item_border etc - this is confusing (and much too verbose)
- an Images API is also badly needed - I hacked up a couple classes but that's exactly that - a hack - see 36e7f8973183dbe16699d8789626663580538fb8. -> new issue #366
- [x] rest of functions in balt to classes (like checkBox, v/hsizer etc)
-  A **single** package with `wx` imports. Single and _as small as possible_ -> **gui**. Files that currently import wx (and they shouldn't) are:
   - [ ] `basher/__init__.py`: 7 usages left (mostly wx.Notebook but also wx.EVT_CONTEXT_MENU,  wx.NOT_FOUND, wx.Platform, wx.EVT_NOTEBOOK_PAGE_CHANGED)
   - [x] `constants.py`: nasty, our settings defaults depended on wx constants
   - [x] `dialogs.py`: sizers, wx.TE_RICH2, wx.ColourPickerCtrl, wx.Icon
   - [x] `frames.py`:
     - [x] this needs attention. wx.lib.iewin --> this must go! Also a lot of classes in need of wrapping
     - [x] wx.ArtProvider_GetBitmap (icons API !)
     - [x] wx.FlexGridSizer
     - [x] One random usage of wx.NOT_FOUND
   - [x] `gui_fomod.py`: same problem as belt below
   - [x] `gui_patchers.py`: sizers, wx.Window
   - [x] `patcher_dialog.py`: sizers, event binding
   - [x] `settings_links.py`: sizers
  And outside basher:
   - [x] `belt.py`: the wizard dialog
   - [x] id parameters from balt functions/constructors must be dropped
   - [x] balt functions such as `button` must become classes and be inherited by OkButton, CancelButton etc

Some newer TODOs:
- [x] where did the tracebacks go? -> fixed in 6f121cb432f13c39d67c07dcabf512d17dcb3ada
- [x] the issue with warnings on booting still allowing the user to interact with main frame -> fixed on wx4 yey! -> https://trac.wxwidgets.org/ticket/17816
- [x] the progress dialog - ugh. Our whole `CancelError` stuff is not working [anymore](https://github.com/wrye-bash/wrye-bash/commit/4766e1051b1e2f534a24b474dfd2cf2c20cf8aec) (and I never liked that exception control flow anyway). ~~More importantly, there is the strange issues of losing focus and I think it's the progress window again.~~ -> wx bug, fixed by switching to GenericProgressDialog -> see #728
- [x] related to the above - make sure we don't refresh twice - we must follow a bit the events (with good old prints) and make sure they are processed once - I am thinking also of cases we display dialogs and such, is RefreshData triggered? More `@balt.conversation` ? We could maybe add an `_is_shown` attribute and use the `evt_activate` argument to only refresh when the `_is_shown` becomes True from False -> seems working ok but keep an eye open!
- [x] not sure about properties yet - sometimes it's ok but sometimes methods are preferred - more explicit - WIP but stabilizing - current code is fine
- [ ] events - it seems wx did not doubly register a handler - see notes on 4766e1051b1e2f534a24b474dfd2cf2c20cf8aec. We do and we skip by default which means we have to go through the previous uses and carefully evaluate if skip is needed. Seems it's not the default because it is not meant to be used often. So there are system handlers that might need to be skipped - could not really understand their docs: https://wxpython.org/Phoenix/docs/html/events_overview.html#how-events-are-processed - I guess system handlers run after user ones
  - [ ] revisit event handling during boot - especially size events need be handled after setting sizes before maximizing apparently etc - see 4d70274ad9a8791723d2e683b7b5467a8809b540 (Attempt to fix installers tab positions not being set)
- [ ] balt will no longer serve a purpose once we're done - all balt code should either go into basher (highly specific Wrye Bash code) or into gui (pure abstractions over wx). Useful rule of thumb: *if you think the class / method could be easily copied into another project without dragging core Bash code along while still being useful, it's gui*
- [ ] revisit locale handling - see> [wxWidgets/Phoenix#1702](https://github.com/wxWidgets/Phoenix/pull/1702) from https://github.com/wrye-bash/wrye-bash/issues/15#issuecomment-690801208

Under #163

**Comment by Utumno** [comment]
To the backburner for now

**Comment by Utumno** [comment]
@barcharcraz - have a look at this one and at the eternal #163 and #174 to get a feeling of how far we are now from the code 4 years ago where basically the wx library was _everywhere_. Still work to do but far from the situation in 054970e7b4d82e351c9a696b8e3c9750516d96b8

**Comment by Utumno** [comment]
@Infernio I have kept a close eye (rather nose cause I am on mobile usually) on your tour-de-force with taking over from @nycz the de-wx stuff. I can't really comment (due to my limited means + the size of the endeavour). Just couple observations:

- `__slots__` are not really needed here (except if I miss something) methinks. I am sure you know all about them so I won't repeat all the blah blah on slots being a memory opt ( https://stackoverflow.com/a/28059785/281545 ), needed when creating millions of instances - in Bash, slots are used in the records code (probably as an optimization or due to the dynamic nature, ahem, of records classes) - this predates me. They are also used by me in the bsa files and save header hierarchies - I used them as kind of internal docs as those are binary files that have a limited and precise number of attributes, so kinda felt natural. They are used in other places (Installer, currently broken, for unpickling - btw MreEffect is also breaking the records slots). So there may be some design/style reasons to use them but those classes are plain old python classes that do need the limitations of slots probably.
- which brings us to "if in doubt leave it out" - great talk on APIs design: https://www.youtube.com/watch?v=heh4OeB9A-c

I like very much the idea of wrapping the events but see talk above - briefly one must use an API to evaluate it.

----

I still believe that we need to first address the webview/iewin as a first step (see . If this locks us to wx 2.9 then high time for a nightly with 2.9 ;)

**Comment by Infernio** [comment]
Yeah, the slots are probably not needed. I threw them in cause I had no idea just how many instances of the GUI classes we create, but it really isn't that many.

**Comment by Infernio** [comment]
Slots are gone now, also cleaned up everything that was commented on.

**Comment by Utumno** [comment]
Thanks! another thing that bothers me btw is my UIList being a wx Panel -
should also be a wrapper of the wx control `__gList` but one step at a time

**Comment by Utumno** [comment]
@Infernio - link to SO question on wx locale -
https://stackoverflow.com/q/54178028/281545

**Comment by Utumno** [comment]
@lojack5 @Infernio I pushed the wx part of lojack-py3 branch on dev - if those changes are wx3 compatible we should (squash) merge those on dev (except if they are contained already in wx-begone branch) - under #460 too

**Comment by Infernio** [comment]
> dunno how I unassigned nycz btw ?

Huh. Normally github does that for people who no longer have a connection to the project. He's still part of the org though, so that's not it: https://github.com/orgs/wrye-bash/people

**Comment by Infernio** [comment]
Left some comments on that branch. Pretty sure only 36471b38f760f16c1c98b75fd20a692e1e6936ce, 5dab3fa60f97d3edc6a0270a5a28949d1c53b448 and 4257102245adfab0169022d389f48d2d640f732f are backwards-compatible, and we need to test if 6bcd48fbcf864c2fc9eb52556fffacf3c9f3f069 is still needed with `inf-wx-begone`.

**Comment by Utumno** [comment]
Thanks! I leave it to you to do the squashing - will turn the eye on wx-begone with your permission will stuff there my couple commits from my records branch -  wx stuff (including the wx-begone one) are our current blocker for py3, plus it's already on nightly for a while, I just want to have a more careful look before merging

**Comment by Infernio** [comment]
> will turn the eye on wx-begone with your permission will stuff there my couple commits from my records branch 

Yup, feel free
We should think about merging (at least part of) wx-begone soon as a `190-de-wx-part-1` merge, since `no-ff`'ing a 30+ commit branch would look pretty silly. I put in quite a bit of effort to make sure none of the commits break dev **as long as they are merged in sequence**, so you could merge e.g. the first 10 commits without breaking dev, but dropping any of those commits *would probably* break dev.

**Comment by Utumno** [comment]
Let's have one last squash on nightly to check my commits also (if you agree with fill -> expand) and then merge time - haven't really reviewed, but I believe the wrappers now are as powerful as wx, with the probable addition of with_border to LayoutOptions (to allow for different borders on top/btm/right/left) - so we're good to go

**Comment by Infernio** [comment]
> if you agree with fill -> expand

Yeah sure, looks fine
Will squash and build one more nightly with those changes

**Comment by Utumno** [comment]
Finally got round to working with the new APIs in wx-begone - must say it's a pleasurable experience :)

We should focus on merging, as much depends on it but I would like some feedback on the commits I (force, hope you don't mind!) pushed on inf-wx-begone - especially the changes on the _AComponent api.

The ones  in utumno-wx-begone-broken are broken (haven't got round to tracing the 'tree' ListBoxes) but show how these new apis could be used to encapsulate some of the tougher nuts, like Dialogs and ListBoxes

**Comment by Infernio** [comment]
Left some preliminary comments on that branch, and here are some thoughts on `utumno-wx-begone-broken`:
> probably this class should be moved to gui/

Yup, my (still pretty far-off...) vision is to have every direct abstraction over wx ugliness in `gui`, leaving us `basher` as a kind-of still pretty WIP 'controller'.
Not sure about balt yet, it might just disappear with its contents split into gui/basher as appropriate? Would mean we get to drop a silly name...
> maybe the time of the ListDataEditor has come

Yes please :) That class is... not the prettiest (see also #484)
> CheckListBox is probably needed instead of stuffing all its methods to ListBox

Definitely, that many methods make the API almost impossible to grasp.

And regarding the whole API for that - I heavily used properties for `_AComponent`, mostly because that's how the code I inherited from nycz worked. But I actually quite like how it looks now - so we should probably make a decision on whether (future?) components should try to use properties first and methods second as well.

**Comment by Utumno** [comment]
Will move new classes to gui/misc_components (?) in next iteration - I agree with the vision:)

> Yes please :) That class is... not the prettiest (see also #484)

Why you hate so much my little closure that can't go out by it-self_ ? :P
Revisiting ListDataEditor is not easy to bin and actually simplifies ListEditor - the real issue is encapsulating the index handling better spend our energy there - and OrderedDict that maps the data model items (sometimes Paths) to their string values displayed by the ListBoxes could be a solution to that

> whether (future?) components should try to use properties first and methods second as well.

I reserve properties for when a property is needed - a function is most explicit (there is some _processing_ going on) - existing properties in _AComponent seem ok for now I would say

> Edit: one more thing,

Done!

**Comment by Infernio** [comment]
> Why you hate so much my little closure that can't go out by it-self_ ? :P

Hmm, I think my brain went 'nope' the first time I saw the `self_ = self` assignment - plus pycharm was screaming warnings at me when I first saw that class - but I'll defer to you here :)

**Comment by Utumno** [comment]
I have done a big rebase of this branch but last minute I saw your fixup -
should I cherry pick and force push?

I split your fixups and tried to make commits self contained for merge and
fixed a load of bugs style or otherwise.
I think you will find lots of interesting stuff - random:

- wizards size not remembered present also on dev
- double binding of refresh?
- some work on the API seems pretty much there

in all looks nice :)

**Comment by Infernio** [comment]
My fixups are just two lines, so feel free to push and cherry-pick them afterwards.

**Comment by Utumno** [comment]
Pushed and pushed again - will be rebasing again soon have a look :)

**Comment by Utumno** [comment]
New iteration - stable enough for nightly. Among others fixes double binding of RefreshData which IIUC was not a problem in the previous implementation but is a problem now. Could write volumes but the rebase exhausted me :P - well, the basic reason I started this was because I needed to use the new APIs to see if they are up to the task. In general, they were a pleasure to use and are quite powerful. Some TODOs:

- I would like to go through the uses of sizers a bit and see what is left out - the wx.TOP etc stuff for instance - I have a plan for adding them in the LayoutOptions
- the progress dialog - ugh. Our whole `CancelError` stuff is not working [anymore](4766e1051b1e2f534a24b474dfd2cf2c20cf8aec) (and I never liked that exception control flow anyway). More importantly, there is the strange issues of losing focus and I think it's the progress window again.
- related to the above - make sure we don't refresh twice - we must follow a bit the events (with good old prints) and make sure they are processed once - I am thinking also of cases we display dialogs and such, is RefreshData triggered? More `@balt.conversation` ? We could maybe add an `_is_shown` attribute and use the `evt_activate` argument to only refresh when the `_is_shown` becomes True from False
- the issue with warnings on booting still allowing the user to interact with main frame
- not sure about properties yet - sometimes it's ok but sometimes methods are preferred - more explicit - WIP
- events - it seems wx did not doubly register a handler - see notes on 4766e1051b1e2f534a24b474dfd2cf2c20cf8aec. We do and we skip by default which means we have to go through the previous uses and carefully evaluate if skip is needed. Seems it's not the default because it is not meant to be used often. So there are system handlers that might need to be skipped - could not really understand their docs: https://wxpython.org/Phoenix/docs/html/events_overview.html#how-events-are-processed - I guess system handlers run after user ones

**Comment by Metallicow** [comment]
@Utumno According to robin...
`Refresh` does not immediately cause a paint event to be sent. It just marks the window as "needing to be painted" and then the system will get around to sending the actual paint event when the time is right.

...Basically this is right, so if you expect Refresh to automatically update.. then that isnt always the case. Note that doing a Refresh then Update usually does. 
Also send size event or PostSizeEvent normally refeshes the painted window with out the user having to grab a gripper(tho, this is tricky when app is fullscreen or minimized) so note that too. Don't want you doing a bit of work only to have you reply "why did it not work like I expected"

**Comment by Utumno** [comment]
@Metallicow thanks I was talking about RefreshData not the wx Refresh

@Infernio wait before new nightly will have to do a couple fixups (and a new rebase)

**Comment by Utumno** [comment]
Done :)

**Comment by Infernio** [comment]
Something in the recent app button refactoring seems to have broken the game button's behavior when a script extender is installed.

Specifically, if no extender is installed and you click the button, the regular launcher should run. If an extender is installed, then there will be a green checkbox in the status bar. If that is not ticked, then the regular launcher should run, same as above.

If it is ticked, then the extender's binary (e.g. skse64_launcher.exe for SKSE64) should directly be run, enabling the extender and skipping the launcher. This last part doesn't seem to work anymore, instead the regular launcher is ran even with the green checkbox ticked.

**Comment by Utumno** [comment]
I see in the source:

            exePath = (self.exePath if not exeObse.exists() and
                       self.exePath.tail == u'Oblivion.exe' else exeObse)

Did this ever work for other games than oblivion? :P Although this means that it always launch **SE for other games

I had a go at replacing splitters and pushed a fixup for those issues too goodnightzzzz

**Comment by Metallicow** [comment]
I don't know because I don't use total conversions or mods like that, but that would be an interesting point to investigate to rewrite that line of code.... It does appear to raise questions the way the logic is written...

**Comment by Utumno** [comment]
belt is broken in this WIP and scratching my head as to why - at some point we should return a PageInstaller from RunLine btw  whose return value is never used

Rebased and pushed it should be ready to test - moved my belt attempts at: ut-belt-xxx branch

**Comment by Utumno** [comment]
Well figured out but prepare for a horrible hack that I moreover can't make to work - testing with


[Kill The Orchestra V3.1-52462-3-1.7z.zip](https://github.com/wrye-bash/wrye-bash/files/4251952/Kill.The.Orchestra.V3.1-52462-3-1.7z.zip)

on skyrim

Turns out we need to *monkey-patch the native widget's GetNext/Prev* - but although works in the debugger does not work in practice showing a Finish button that works as Next however

Have a look on my branchy ut-belt-xxx

**Comment by Utumno** [comment]
@Infernio on next nightly iteration please squash current 190-wx-begone-pt1 as it contains lots of fixups

**Comment by Infernio** [comment]
I figured out the cause of the xSE issue I reported above 😅:
```diff
diff --git a/Mopy/bash/basher/app_buttons.py b/Mopy/bash/basher/app_buttons.py
index 5894b78af..18a50feff 100644
--- a/Mopy/bash/basher/app_buttons.py
+++ b/Mopy/bash/basher/app_buttons.py
@@ -495,7 +495,7 @@ def Execute(self):
             BashFrame.modList.RefreshUI(refreshSaves=True, focus_list=False)

 #------------------------------------------------------------------------------
-class Game_Button(_App_Button):
+class Game_Button(_ExeButton):
     """Will close app on execute if autoquit is on."""
     def __init__(self, exe_path_args, version_path, images, tip, obse_tip):
         exePath, exeArgs = _parse_button_arguments(exe_path_args)
```

**Comment by Infernio** [comment]
Re:
> where did the tracebacks go?

If you mean the popup windows from wxPython that used to appear, they broke at some point - probably during 050391ca22d7c8451390cbff6fd150ab0b9bcabd. Haven't been able to figure out how to get them back yet though :/

**Comment by Infernio** [comment]
Ah, I think managed to get them back. It happens because the default for redirecting seems to have been changed in wx3, now it *doesn't* redirect by default.

**Comment by Utumno** [comment]
Thanks - yep that app button code definitely needed some dusting turns out. I worked a bit on ListBoxes, replaced the VLayout with GridLayout to mimic the old behavior a bit. Good news - we can run at wx4 have a look at lojack-py3-wx-dev

**Comment by Infernio** [comment]
> Good news - we can run at wx4 have a look at lojack-py3-wx-dev

Ah, nice - I tried porting a few weeks ago and it didn't seem overly difficult, will take a look :)

**Comment by Infernio** [comment]
Hiding status bar buttons has stopped working, it fails because of this check:
https://github.com/wrye-bash/wrye-bash/blob/99eac1491742985057f94e4663abbd20064e52ba/Mopy/bash/basher/__init__.py#L3607

`self.buttons` contains wrapped classes, but `self.window` for status bar links seems to always be a wx object. We could make the status bar directly work with wx objects again - but making it *not* do that was kind of the point :/

**Comment by Infernio** [comment]
A random thought: One thing we haven't tried with progress dialogs is setting the parent correctly. We almost always set it to `None` (since that's our default value), but e.g. mods_metadata sets it to the mod checker instance it's working one.

**Comment by Infernio** [comment]
@Utumno ListBoxes still don't seem to work right, the 'children affected' warning when you disable a mod is collapsed by default and expands really weirdly:

![ListBoxes](https://user-images.githubusercontent.com/2845240/75613542-2fb1e680-5b2f-11ea-896b-456cbcbc0d37.png)

![ListBoxes2](https://user-images.githubusercontent.com/2845240/75613543-32acd700-5b2f-11ea-9d77-66004114f731.png)

**Comment by Metallicow** [comment]
That isn't a picture of a listbox, that is a treectrl of some sort or another. You might need to send a expand function after widget is created in your wrapper code.

The expansion you are seeing is almost certainly what flags args you set up for the sizers.
Ex: sizer.Add(widget, normally 1 or 0(1 will expand with flag), flags like wx.EXPAND, border)
Turn the statictext and the okbutton to 0 and/or remove a flag.

**Comment by Utumno** [comment]
Here is what I see:

![tree-tesiv](https://user-images.githubusercontent.com/1365079/75624445-d4c6d080-5bbc-11ea-8002-9f8028b4fb46.jpg)

And on clicking the +

![tree-tesiv-expand](https://user-images.githubusercontent.com/1365079/75624450-e60fdd00-5bbc-11ea-8334-42e7061d2f68.jpg)

We need to checkout the wx2 versions at some point and take it from there

Thanks for mass_select wrappers :) I saw a note on stars in espm - that is for renamed plugins IIRC, haven't used it much

**Comment by Utumno** [comment]
Pushed another rebase - getting there. Note I dropped some edits whitespace in some translatable strings - we want to avoid this as translations break and this is not a good idea - that's 308 anyway.

Meanwhile a few things I would like to work on:

- more list boxes would like to have the actual list expand to full width - max size I considered but  adds unneeded complexity, plus in general I am fun of arbitrary resize (and revered apps like git gui allow it too). Also, the "extend by default might lead to some issues:

![tree-expand](https://user-images.githubusercontent.com/1365079/75630245-9d264b80-5bf1-11ea-9b09-dff28d0f8cca.jpg)

:p

- turn some properties to methods - in particular, tooltip behaves not so well - it disappears and one can't get it easily back - probably a private tooltip field and then some show/hide_tooltip but not trivial - anyway this is not a property for sure. background color and visible too probably

for pt2 I'd say the image handling (balt.Picture) is a good first candidate, it's all over and last couple uses of setup_hook are for this (and the UILIst edit control which needs more work however before pt1)

Thanks for carrying the refactoring to fomods! And thanks for the lz4 merge, oh dear, we can fibnally deconmpress arbitrary ba2, that always bugged me

**Comment by Infernio** [comment]
Yeah, I noticed that... problem is if you don't expand them by default, they remain extremely tiny and need to be resized every time. We could maybe remember the size and reapply it?

**Comment by Utumno** [comment]
Hmm - need to have a look at how this used to be handled - probably set a min size to the TreeCtrl would be  asimpler solution - ListBox ctrls display alright it seems

**Comment by Infernio** [comment]
*(also, please ignore the issue spam, I'm reworking our tags a bit with a system I totally didn't steal from Rust's repo)*

**Comment by Metallicow** [comment]
@Infernio Here is a very basic sample app. Note the proportion value with EXPAND flag.

<details>
      <summary>SampleApp - Click to expand</summary>

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Imports.--------------------------------------------------------------------
# -Python Imports.
import os
import sys

# -wxPython Imports.
import wx

class MyPanel(wx.Panel):
    def __init__(self, parent, id=wx.ID_ANY,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.BORDER_SUNKEN, name='panel'):
        wx.Panel.__init__(self, parent, id, pos, size, style, name)

        st = wx.StaticText(self, -1, label='testing statictext')
        self.tc = wx.TreeCtrl(self, -1)
        self.rootitem = self.tc.AddRoot('Root')
        for modchildindex in range(7):
            treeitemid = self.tc.AppendItem(self.rootitem, 'item %d' %modchildindex)
        self.tc.ExpandAllChildren(self.rootitem)
        self.okbtn = wx.Button(self, wx.ID_OK)

        border = 5
        vbSizer = wx.BoxSizer(wx.VERTICAL)
        vbSizer.Add(st, 0, wx.EXPAND | wx.ALL, border)
        vbSizer.Add(self.tc, 1, wx.EXPAND | wx.ALL, border)
        vbSizer.Add(self.okbtn, 0, wx.EXPAND | wx.ALL, border)
        self.SetSizer(vbSizer)


class MyFrame(wx.Frame):
    def __init__(self, parent, id=wx.ID_ANY, title=wx.EmptyString,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.DEFAULT_FRAME_STYLE, name='frame'):
        wx.Frame.__init__(self, parent, id, title, pos, size, style, name)
        global gMainWin
        gMainWin = self
        wxVER = 'wxPython %s' % wx.version()
        pyVER = 'python %d.%d.%d.%s' % sys.version_info[0:4]
        versionInfos = '%s %s' % (wxVER, pyVER)
        self.CreateStatusBar().SetStatusText(versionInfos)
        panel = MyPanel(self)
        self.Bind(wx.EVT_CLOSE, self.OnDestroy)

    def OnDestroy(self, event):
        self.Destroy()

class MyApp(wx.App):
    def OnInit(self):
        self.SetClassName('MyApp')
        self.SetAppName('MyApp')
        gMainWin = MyFrame(None)
        gMainWin.SetTitle('MyFrame')
        self.SetTopWindow(gMainWin)
        gMainWin.Center()
        gMainWin.Show()
        return True

if __name__ == '__main__':
    gApp = MyApp(redirect=False,
                 filename=None,
                 useBestVisual=False,
                 clearSigInt=True)
    gApp.MainLoop()

```
</details>

**Comment by Metallicow** [comment]
@Utumno Using the wx.RESIZE_BORDER flag in construction of the frame/dialog allows griper and border to be resizable.

**Comment by Utumno** [comment]
> sometimes SetSizerAndFit locks the frame/dialog minsize also, so that may be the problem also.

Well turns out this was the issue - current version ListBoxes are better than ever (or at least than current dev). Will squash all this soon

**Comment by Utumno** [comment]
Just to say I deleted the lojack-py3-wx-dev as is absorbed into 488-wx4-upgrade and I updated the 190-de-wx with more fixups (and squashing)

Deleted also ganda-460-part1 as is in nightly

**Comment by Utumno** [comment]
Slight heads up - I am performing yet another rebase - tried finalizing UIListCtrl merge but ended up in the perilous water of its event handling so this will take a while. If you commit anything please commit it on top of the current branch for easier merging - don't rebase yet please.

Some difficulties:

- in this event handling code event.Skip() is *not* usually called - this may be on purpose but we should double-check - an opportunity to get this right (and add comments) - cf comment on Veto  in this code which *are* needed - related to exceptions thrown on renaming. Testing all possible paths will take a while
- `bak` savefiles rename is blocked currently. This necessitates an override validate_filename which must _go_. So how should those bak be renamed?
- technicalities that need testing for instance event.GetText() vs event.GetLabel() etc (docs are rather unhelpful https://wxpython.org/Phoenix/docs/html/wx.ListEvent.html#wx.ListEvent.GetText )

**Comment by Infernio** [comment]
I won't be around to do anything for a while (exams), so don't worry about me rebasing anything :P

**Comment by Utumno** [comment]
All the best with exams!

Posted new rebase - fixing a few bugs also. When you get round to it time for a new nightly?

**Comment by Utumno** [comment]
Thanks for fixup there were errr - a couple more - including a fix for cosaves refresh. Getting there (have tried quite a few things, some of it is on the branch - this playing with boot made me discover the saves refresh issue)

**Comment by Infernio** [comment]
@Utumno There's a report on discord that using Ctrl+A to select all mods doesn't work anymore (no traceback for it, just seems to silently fail).
Also a report that changing status bar icon sizes crashes - last I investigated, changing status bar size is impossible in wxPython, so maybe we should just drop that feature completely if it can't be repaired?

**Comment by Utumno** [comment]
re ctrl+a I will test asap

re: icons - maybe destroying and recreating the status bar? or drop if
can't be fixed yes

**Comment by Metallicow** [comment]
SetBitmap, and Refresh then Update on the parent window should work for changing a bitmap size. You may have to call layout on the sizer depending on the os. or just call PostSizeEvent on the mainwin afterwards.
If you add/inject a PyShell/Crust into the app, you should be able to live debug that one since the source code is already sitting right in your lap.
Tho handling the int to be saved as user pref would be usually before or after the change in bitmap size. Ex: 16, 24, 32 so that might get screwed up if live debuging if you forget to update that value.

**Comment by Utumno** [comment]
Spent some time over the status bar - there are various subtly broken things and my attempt at wrapping went not so good - Ctrl+A seems ok on current 190-de-wx

**Comment by Utumno** [comment]
@Infernio see my post above - and yes disabling them icon sizes seems the best solution, I tried recreating the satus bar but this also hard crashes:

```

------------------------- Mopy/bash/basher/__init__.py -------------------------
index dad02dd90..22785fbd2 100644
@@ -3672,3 +3672,3 @@ def _do_refresh(self, refresh_icon_size=False):
         self.SendSizeEventToParent()
-        self.OnSize()
+        if not refresh_icon_size: self.OnSize()
 
@@ -3706,3 +3706,3 @@ def __init__(self, parent=None):
         #--Status Bar
-        self._native_widget.SetStatusBar(BashStatusBar(self._native_widget))
+        self.frame_set_sb()
         #--Notebook panel
@@ -3719,2 +3719,8 @@ def __init__(self, parent=None):
 
+    def frame_set_sb(self):
+        if self._native_widget.GetStatusBar():
+            self._native_widget.GetStatusBar().Hide()
+            self._native_widget.GetStatusBar().Destroy()
+        self._native_widget.SetStatusBar(BashStatusBar(self._native_widget))
+
     @balt.conversation

---------------------- Mopy/bash/basher/settings_links.py ----------------------
index ea564096e..4fd770395 100644
@@ -254,3 +254,4 @@ def Execute(self):
         bass.settings['bash.statusbar.iconSize'] = self.sb_icon_size
-        Link.Frame.statusBar.UpdateIconSizes()
+        # Link.Frame.statusBar.UpdateIconSizes()
+        Link.Frame.frame_set_sb()
 
```


Thanks for the nightly! that's pretty much stable apart from some fixes for the status bar drag and drop - contained in newest 190-de-wx :)

**Comment by Utumno** [comment]
Script for testing status bar size stuff - a hasty start - size seems being reset no matter what on resizing the frame, and never looks ok

```python
import wx

class MainFrame(wx.Frame):

    def __init__(self):
        wx.Frame.__init__(self, None, title='Statusbars')
        panel = wx.Panel(self)
        status_btn = wx.Button(panel, label='Set status bar size')
        status_btn.Bind(wx.EVT_BUTTON, self.change_size)
        restore_btn = wx.Button(panel, label='Print status bar size')
        restore_btn.Bind(wx.EVT_BUTTON, self.print_size)
        self.txtctrl = wx.TextCtrl(panel)
        self.txtctrl.Bind(wx.EVT_TEXT, self.set_size)
        # self.txtctrl.Bind(wx.EVT_KILL_FOCUS, self.set_size)
        sizer = wx.BoxSizer(wx.HORIZONTAL)
        sizer.Add(status_btn, 0, wx.ALL, 5)
        sizer.Add(restore_btn, 0, wx.ALL, 5)
        sizer.Add(self.txtctrl, 0, wx.ALL, 5)
        panel.SetSizer(sizer)
        self.statusbar = self.CreateStatusBar(2)
        self.statusbar.SetStatusText('This goes in field one')
        self.statusbar.SetStatusText('Field 2 here!', 1)
        self._sb_size = 32
        self.Show()

    def change_size(self, event):
        print (self.statusbar.GetSize())
        self.statusbar.SetSize((-1, self._sb_size))
        print (self.statusbar.GetSize())

    def print_size(self, event):
        self.statusbar.SetStatusText('Size %s' % self.statusbar.GetSize())
        self.statusbar.Update()
        self.statusbar.Refresh()

    def set_size(self, event):
        try:
            self._sb_size = int(self.txtctrl.GetValue())
        except ValueError as e:
            pass #print (e)

if __name__ == '__main__':
    app = wx.App(False)
    frame = MainFrame()
    app.MainLoop()

```

**Comment by Metallicow** [comment]
@Utumno This might help make it a bit clearer when updating the statusbar. It isn't the prettiest way to write it, but it is basically how it works. Of course you will have to update the bitmap sizes also, as I haven't provided a image for this sample.

<details>
      <summary>StatusBar AdjustSize/Height Sample - Click to expand</summary>

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""""""

# Imports.--------------------------------------------------------------------
# -Python Imports.
import os
import sys

# -wxPython Imports.
import wx

class MyStatusBar(wx.StatusBar):
    def __init__(self, parent, id=wx.ID_ANY, style=0, name='statusbar'):
        style = (wx.STB_DEFAULT_STYLE
                 # | wx.STB_SIZEGRIP
                 # | wx.STB_ELLIPSIZE_END
                 # | wx.STB_ELLIPSIZE_MIDDLE
                 # | wx.STB_ELLIPSIZE_START
                 # | wx.STB_SHOW_TIPS
                 )
        wx.StatusBar.__init__(self, parent, id, style, name)

        self.btn1 = wx.BitmapButton(self, -1, wx.Bitmap('phoenix32.png'))
        self.SetMinSize((-1, 32 + 10))

        self.Bind(wx.EVT_RIGHT_UP, self.OnRightUp)

    def OnRightUp(self, event):
        menu = wx.Menu()
        for heightinpx in (16, 24, 32):
            newid = wx.NewId()
            h = u'%d' % heightinpx
            menuitem = wx.MenuItem(menu, newid, h, h)
            menu.Append(menuitem)
            menu.Bind(wx.EVT_MENU, self.OnSetStatusBarHeight, id=newid)
        self.PopupMenu(menu)
        menu.Destroy()

    def OnSetStatusBarHeight(self, event):
        """
        Set the statusbar height from menuitem label.

        :param `event`: A `wx.MenuEvent` to be processed.
        :type `event`: `wx.MenuEvent`
        """
        self.SetMinHeight(int(event.GetEventObject().GetLabel(event.GetId())) + 10)
        self.SetMinSize((-1, int(event.GetEventObject().GetLabel(event.GetId())) + 10))
        self.SetSize((-1, int(event.GetEventObject().GetLabel(event.GetId())) + 10))
        self.Refresh()
        self.Update()
        self.PostSizeEvent()
        # wx.CallAfter(gMainWin.Refresh)
        # wx.CallAfter(gMainWin.Update)
        wx.CallAfter(gMainWin.PostSizeEvent)

class MyPanel(wx.Panel):
    def __init__(self, parent, id=wx.ID_ANY,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.BORDER_SUNKEN, name='panel'):
        wx.Panel.__init__(self, parent, id, pos, size, style, name)

class MyFrame(wx.Frame):
    def __init__(self, parent, id=wx.ID_ANY, title=wx.EmptyString,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.DEFAULT_FRAME_STYLE, name='frame'):
        wx.Frame.__init__(self, parent, id, title, pos, size, style, name)
        global gMainWin
        gMainWin = self
        wxVER = 'wxPython %s' % wx.version()
        pyVER = 'python %d.%d.%d.%s' % sys.version_info[0:4]
        versionInfos = '%s %s' % (wxVER, pyVER)
        self.statusbar = MyStatusBar(self)
        self.SetStatusBar(self.statusbar)
        self.statusbar.SetStatusText(versionInfos)
        panel = MyPanel(self)
        self.Bind(wx.EVT_CLOSE, self.OnDestroy)

    def OnDestroy(self, event):
        self.Destroy()

class MyApp(wx.App):
    def OnInit(self):
        self.SetClassName('MyApp')
        self.SetAppName('MyApp')
        gMainWin = MyFrame(None)
        gMainWin.SetTitle('MyFrame')
        self.SetTopWindow(gMainWin)
        gMainWin.Center()
        gMainWin.Show()
        return True

if __name__ == '__main__':
    gApp = MyApp(redirect=False,
                 filename=None,
                 useBestVisual=False,
                 clearSigInt=True)
    gApp.MainLoop()

```
</details>

**Comment by Infernio** [comment]
Got this upon closing (with X, not OK) the color dialog:
```
Traceback (most recent call last):
  File "bash\gui\events.pyo", line 168, in _post
  File "bash\basher\dialogs.pyo", line 176, in OnOK
  File "bash\basher\dialogs.pyo", line 251, in on_closing
  File "bash\gui\misc_components.pyo", line 117, in unsubscribe_handler_
  File "bash\gui\events.pyo", line 208, in unsubscribe
bash.exception.UnknownListener: Listener <bound method DropDown._set_tooltip of <bash.gui.misc_components.DropDown object at 0x000000001339CF48>> not subscribed on EventHandler(<wx._core.ComboBox object at 0x00000000133871F8>, <wx.core.PyEventBinder object at 0x0000000004B8FC08>)
```
Will investigate later, basically just posting this here as a reminder to myself :)

**Comment by Infernio** [comment]
Looks like we hit `ColorDialog.on_closing` twice, once from a direct callback and once from `ColorDialog.OnOk` (which I guess the modal dialog implicitly calls?). We could unsubscribe the `on_closing` handler in `ColorDialog.__init__`, but I don't really like that either...

**Comment by Utumno** [comment]
See 3bf5c7b67cd7bfb7b889c7a34f2dabbc56cafdde search commit msg for UnknownListener - thought I fixed it but not quite

**Comment by Utumno** [comment]
Tweaked a bit the script by Metallicow above

<details>
      <summary>StatusBar AdjustSize/Height Sample - Click to expand</summary>

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import sys
import wx
imdir = r"C:\Dropbox\eclipse_workspaces\python\wrye-bash\Mopy\bash\images"
pjoin = lambda x: os.path.join(imdir, x)
class MyStatusBar(wx.StatusBar):
    def __init__(self, parent, id=wx.ID_ANY, style=0, name='statusbar'):
        style = (wx.STB_DEFAULT_STYLE  # | wx.STB_SIZEGRIP | wx.STB_ELLIPSIZE_END |
                 # wx.STB_ELLIPSIZE_MIDDLE | wx.STB_ELLIPSIZE_START |
                 # wx.STB_SHOW_TIPS
                 )
        wx.StatusBar.__init__(self, parent, id, style, name)
        self._pngs = {32: pjoin(r"checkbox_blue_off_32.png"),
                      24: pjoin(r"checkbox_blue_off_24.png"),
                      16: pjoin(r"checkbox_blue_off.png"), }
        self.iconsSize = 32
        self.btn1 = wx.BitmapButton(self, -1, wx.Bitmap(self._pngs[self.iconsSize]))
        self.SetMinSize((-1, self.iconsSize + 10))
        self.Bind(wx.EVT_RIGHT_UP, self.OnRightUp)

    def OnRightUp(self, event):
        menu = wx.Menu()
        for heightinpx in (16, 24, 32):
            newid = wx.NewId()
            h = u'%d' % heightinpx
            menuitem = wx.MenuItem(menu, newid, h, h)
            menu.Append(menuitem)
            menu.Bind(wx.EVT_MENU, self.OnSetStatusBarHeight, id=newid)
        self.PopupMenu(menu)
        menu.Destroy()

    def OnSetStatusBarHeight(self, event):
        """
        Set the statusbar height from menuitem label.

        :param `event`: A `wx.MenuEvent` to be processed.
        :type `event`: `wx.MenuEvent`
        """
        self.iconsSize = int(event.GetEventObject().GetLabel(event.GetId()))
        self.btn1.Destroy()
        self.btn1 = wx.BitmapButton(self, -1,
                                    wx.Bitmap(self._pngs[self.iconsSize]))
        self.SetMinHeight(self.iconsSize + 10)
        self.SetMinSize((-1, self.iconsSize + 10))
        self.SetSize((-1, self.iconsSize + 10))
        self.Refresh()
        self.Update()
        self.OnSize()
        self.PostSizeEvent()
        # wx.CallAfter(gMainWin.Refresh)
        # wx.CallAfter(gMainWin.Update)
        wx.CallAfter(gMainWin.PostSizeEvent)

    def OnSize(self, event=None):
        rect = self.GetFieldRect(0)
        (xPos, yPos) = (rect.x + 4, rect.y + 2)
        for button in [self.btn1]:
            button.component_position = (xPos, yPos)
            xPos += self.iconsSize
        if event: event.Skip()

class MyPanel(wx.Panel):
    def __init__(self, parent, id=wx.ID_ANY,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.BORDER_SUNKEN, name='panel'):
        wx.Panel.__init__(self, parent, id, pos, size, style, name)

class MyFrame(wx.Frame):
    def __init__(self, parent, id=wx.ID_ANY, title=wx.EmptyString,
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.DEFAULT_FRAME_STYLE, name='frame'):
        wx.Frame.__init__(self, parent, id, title, pos, size, style, name)
        global gMainWin
        gMainWin = self
        wxVER = 'wxPython %s' % wx.version()
        pyVER = 'python %d.%d.%d.%s' % sys.version_info[0:4]
        versionInfos = '%s %s' % (wxVER, pyVER)
        self.statusbar = MyStatusBar(self)
        self.SetStatusBar(self.statusbar)
        self.statusbar.SetStatusText(versionInfos)
        # panel = MyPanel(self)
        self.Bind(wx.EVT_CLOSE, self.OnDestroy)

    def OnDestroy(self, event):
        self.Destroy()

class MyApp(wx.App):
    def OnInit(self):
        self.SetClassName('MyApp')
        self.SetAppName('MyApp')
        gMainWin = MyFrame(None)
        gMainWin.SetTitle('MyFrame')
        self.SetTopWindow(gMainWin)
        gMainWin.Center()
        gMainWin.Show()
        return True

if __name__ == '__main__':
    gApp = MyApp(redirect=False,
                 filename=None,
                 useBestVisual=False,
                 clearSigInt=True)
    gApp.MainLoop()
```
</details>

seems to be able to add the image in the sb and change its size (wx3)

**Comment by Metallicow** [comment]
When it comes to maximized and minimized, back in ancient 2.8 or qt4 you would have to do this ugly hack to force refresh the window. In phoenix there are proper ways to code it that handle this now, but every once and a while you might see this type of crap that someone has put in the code. It is also hard to find because usually coders dont put a comment as to why they did this....
```python
def ForceHackRefresh():
    """Ugly hack to force refresh window when max/min imized"""
    x, y = self.GetSize()
    self.SetSize((x + 1, y))
    self.SetSize((x - 1, y))
```

**Comment by Infernio** [comment]
Report on Discord:

> Hello. I've encountered a weird issue with the current WIP build when using a wizard. I was trying to build a wizard for a mod that only creates an Ini tweak file (similar to that mod's OBMM script). It works fine in the current release, but in the WIP build I get an error at the end 
>
> ```
> Traceback (most recent call last):
>  File "bash\balt.pyo", line 2019, in __Execute
>  File "bash\balt.pyo", line 1134, in _conversation_wrapper
>  File "bash\basher\installer_links.pyo", line 265, in Execute
>  File "bash\gui\misc_components.pyo", line 287, in lb_check_at_index
> wx._core.wxAssertionError: C++ assertion "IsValid(uiIndex)" failed at ....\src\msw\checklst.cpp(276) in wxCheckListBox::Check(): bad wxCheckListBox index
> ```
>
> Seems to be because this mod has no subfolders (there are no optional files for install). The error went away when I created two dirctories in the installer, one dummy directory and one that contained the esp.

**Comment by Metallicow** [comment]
looks to be just basic logic error in wxCheckListBox. It is probably right. See your edit.

**Comment by Infernio** [comment]
Another one from Discord, looks like we're feeding a value that is higher than its maximum to a `balt.Progress` at some point in BAIN?

>Got this popup when opening Oblivion Installers, then tried in debug mod and got the same result, I cannot reproduce it now though. From BashBugDump:
>
>```
>Traceback (most recent call last):
>  File "bash\basher\__init__.pyo", line 3560, in OnShowPage
>  File "bash\balt.pyo", line 1134, in _conversation_wrapper
>  File "bash\basher\__init__.pyo", line 2913, in ShowPanel
>  File "bash\balt.pyo", line 1134, in _conversation_wrapper
>  File "bash\bosh\bain.pyo", line 1559, in _projects_walk_cache_wrapper
>  File "bash\basher\__init__.pyo", line 2943, in _refresh_installers_if_needed
>  File "bash\bosh\bain.pyo", line 1644, in irefresh
>  File "bash\bosh\bain.pyo", line 1995, in _refresh_from_data_dir
>  File "bash\bosh\bain.pyo", line 131, in final_update
>  File "bash\bosh\bain.pyo", line 158, in calc_crcs
>  File "bash\bolt.pyo", line 1901, in __call__
>  File "bash\bolt.pyo", line 1869, in __call__
>  File "bash\balt.pyo", line 1093, in _do_progress
>wx._core.wxAssertionError: C++ assertion "value <= m_maximum" failed at ..\..\src\msw\progdlg.cpp(369) in wxProgressDialog::Update(): invalid progress value
>```
>
> Not sure if it means anything, it did not prevent me from doing things.

**Comment by Infernio** [comment]
@Utumno Done - kept some commits (e.g. 597c6a619483054eaffbf2cf426997490138282b and b10aaf03605965afeb43ad8b118351a883a90496) separate, since they modify the GUI layout instead of just refactoring it. Some others could be merged into random nearby commits (e.g. d4e911e1add3865534791e0d4062a1ffb0cb50e0 and bc13316a34afb8a65c6100fe056f73a8eb8b259d).
I also made ba2bd0aa078456c2db83f34c94238be849ae7d87 no longer spit out #190 pings (see above) - be sure to undo that again when merging :)

**Comment by Utumno** [comment]
Thanks! I think I will leave those commits separate but still have to work a bit on my messages and some random noop edits

**Comment by Utumno** [comment]
Done :)

I left the merge commit message for you to edit - then let's move on

Yey!

**Comment by Infernio** [comment]
Done 🎉🎉🎉

**Comment by Utumno** [comment]
Wow :champagne:

I can't say how much relieved I feel about this, was a huge blocker

I was going through nightly to see which commits we should move to dev - I think the next big step would be deciding on some py3 handling - especially strings and `from future`. Do we really want to add 25k u prefixes that make lines longer and will soon look ugly? Do we really want to add from future just to drop them afterward?

**Comment by Infernio** [comment]
I merged a few of Ganda's commits from nightly already, since a lot of other commits depend on them so we can't merge those unless Ganda's commits are merged - see 33eac7624971ecd22e1f65ff5e47bc71ca175dbc.

Beyond that though - my opinion is that we *should* do the whole prefixing thing once and then let 2to3 drop them again. Yes, this will leave us with an extremely ugly commit in the history that modifies almost every line, but after that we won't ever have to worry about 'needing to prefix/unprefix lines in the vicinity of other edits'. So I say we do this once, then it's over with :)

**Comment by Utumno** [comment]
Ok d'accord, will go through the prefixing in a bit - meanwhile you could add to dev some more commits from nightly? For instance (some of them maybe squashed but most of them on their own)

8dd4fc03d080f808a663ac58b73de12a7f3fc65f
5af752eb6cb482ce7ec1850c65358769952a87cd
c9d9a4100926b190231e8454d2fa83e0c20b361e
e158ae6a02a7167625c4de389a5cdac82bfbd570
f5e82618c950c9afe96e2b6a8a8383e9a768358f
d820d28b6aeb36f71067811d401fdbcf3410e40e
8acc43d8a6b236dd29a145d81e44d52912192002 (SSS?)
4195594adcfe1a7e43010c12da9bb78739e17fb8
...
65dadb3fc85254f0ebf13d04516d417872c4c719  (SSS?)

Haven't thoroughly reviewed but I guess most of them are already on nightly for a while

**Comment by Infernio** [comment]
Right, was just waiting for us to reach a decision :P
If we are going to do the prefixing, then we should try to empty nightly - maybe keep fomod/VR/Morrowind on there? - but we should try to merge as much of it as possible, otherwise merging the prefixing commits will be nothing short of a nightmare.

**Comment by Utumno** [comment]
I do agree - I was thinking however that we should go in steps like settings keys, then dirs etc rather than on a file by file basis. The settings one is especially important and we may need some backwards compat code - these are stored in user files. On the other hand, internal stuff like dirs keys could stay as-is. We will have to try a py2 to 3 converter to see what options are offered.

On the other hand - maybe just revert our handling of byte header signatures? Apart from some wx keyword arguments that probably are fixed in wx4 these are the only bytes we use. This could save us from prefixing all the records files

**Comment by Infernio** [comment]
Re: the byte header signatures - I'm afraid that converting them to unicode strings on the fly will be too expensive.
Consider `parsers.ModHeaderReader.read_mod_headers`. In SSE/FO4, this runs on potentially *millions* of records at the first startup.

Another reason why it may not actually be that bad is that a lot of the bytestrings in records are 100% predictable. Consider 8d99a99d1196e6da61be8bea14027aae74e56609 for example - one regex replace, and all `classType`s are handled. We can do the same for all `MelStruct('XXXX', ...` usages, etc.

**Comment by Infernio** [comment]
@Utumno Dialogs seem to be crashing (tested only on `Create New Project` and `Colors` for now, but it seems like every dialog that doesn't specify a `sizes_dict` is affected):
```
Traceback (most recent call last):
  File "bash\balt.py", line 1792, in __Execute
    self.Execute()
  File "bash\balt.py", line 911, in _conversation_wrapper
    return func(*args, **kwargs)
  File "bash\basher\installers_links.py", line 573, in Execute
    def Execute(self): CreateNewProject.display_dialog(self.window)
  File "bash\gui\top_level_windows.py", line 165, in display_dialog
    with cls(*args, **kwargs) as dialog:
  File "bash\basher\dialogs.py", line 347, in __init__
    super(CreateNewProject, self).__init__(parent)
  File "bash\gui\top_level_windows.py", line 151, in __init__
    sizes_dict, icon_bundle, title=self.title, style=style, **kwargs)
  File "bash\gui\top_level_windows.py", line 48, in __init__
    self._set_pos_size(kwargs, sizes_dict)
  File "bash\gui\top_level_windows.py", line 56, in _set_pos_size
    kwargs['pos'] = kwargs.get('pos', None) or sizes_dict.get(
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by Utumno** [comment]
Oops - regression fix on regression fix let me push

**Comment by Infernio** [comment]
@Utumno Seems like the BP dialog no longer remembers its size? I tried deleting my Oblivion settings files and noticed that I can resize it, but it always goes back to default when I go to rebuild it.

**Comment by Utumno** [comment]
Ooops a regression in the regression fix? Will refix and drop on dev but meanwhile any clue how we could resize the ListBox to be the size of its buttons at least or title (as is now) whichever is bigger? Cause this:

![zzz](https://user-images.githubusercontent.com/1365079/78026298-0d94c980-7364-11ea-828b-ac574bc5897d.jpg)

is getting old

**Comment by Infernio** [comment]
Not a clue, `balt.ListBoxes` is just a weird mess to me :/

**Comment by Infernio** [comment]
Whelp, at least now I know why none of my attempts to get `wx.SplitterWindow` to put a line between the Save/Cancel buttons and the Bash Tags field worked. See Robin Dunn's post [here](http://wxpython-users.1045709.n5.nabble.com/wx-SplitterWindow-styles-doesn-t-work-td2368943.html):
> At about the time that XP was released the wx.SplitterWindow was changed
> to use the system theme for drawing the sash.  So now each platform
> draws it differently but each is consistent with native styles.  On
> Windows that means that the sash itself will not have any borders and
> will rely on the widgets on either side to have some sort of border to
> help show the sash's position.

**Comment by Infernio** [comment]
Not sure if you already figured this out, but just in case:
Remember when we had the issue with one of the dialogs unsubscribing the combobox resize event twice? The reason for that is that if a dialog has a `CancelButton` with a listener subscribed to its click event, wxPython will automatically send a click event for the cancel button when the 'X' in top right corner is clicked. So there is no need for special 'on close' logic as long as a `CancelButton` (i.e. a button with ID_CANCEL) is present on the dialog and does the cleanup.

**Comment by Utumno** [comment]
> So there is no need for special 'on close' logic as long as a CancelButton (i.e. a button with ID_CANCEL) is present on the dialog and does the cleanup.

Yep! Had seen the event fired and went a bit back and forth at some point - was it the color dialog?

**Comment by Infernio** [comment]
Color dialog sounds about right. Only posting this now because I ran into the same problem with the new settings dialog a few days ago.

**Comment by Utumno** [comment]
> For reference's sake: the locale situation seems to maybe have finally gotten better in wx4.1: [wxWidgets/Phoenix#1702](https://github.com/wxWidgets/Phoenix/pull/1702)

Oh that is good news indeed - edited OP

**Comment by Infernio** [comment]
@Utumno I think I've figured out the weirdness where sometimes key codes don't match up. For `EVT_CHAR` events, the key code is what you'd get if the input was inside a text input of some kind. So let's say I press the A key on my keyboard and capture an `EVT_CHAR`:
 - If no modifier keys are pressed, the event will have key code `ord('a') == 97`.
 - If the Shift key is pressed, the event will have key code `ord('A') == 65`.
 - If the Ctrl key is pressed, the event will have key code `_wx.WXK_CONTROL_A == 1`.

The other events like `EVT_KEY_UP` do not do this. I've switched our one usage of `EVT_CHAR` to `EVT_KEY_UP` in my WIP branch, but I'll cherry pick that out and put it on nightly.

This is all documented here: https://wxpython.org/Phoenix/docs/html/wx.KeyEvent.html

**Comment by Infernio** [comment]
The FindString issue might have been this: https://github.com/wxWidgets/Phoenix/issues/1273

**Comment by Utumno** [comment]
I am working on this and will hopefully close it for 312, as it's been around for a while :P and keeps blocking some nice refactoring - some of it was ideas that I had for a while - encapsulate the JIT behavior of some of the balt/gui classes is the `Lazy` class - wrap Objects above Window another one. Now this makes controlling access to the `_native_widget` more imperative, think also of Bash ctd'ing very very hard when for instance wx.App is not initialized (or...?). Think also of switching gui backend, even if we won't ever I hope, a gui wrapper would profit from being designed with this in mind. I was thinking of something like:

```python
+def _native(*, native_get: str = None, native_set: str = None):
+    """Decorator to delegate method calls to the native widget. Used to
+    control native widgets access (and potentially swap gui backend)."""
+    if not (bool(native_set) ^ bool(native_get)): # xor them!
+        raise ValueError(f'Pass either {native_get=} or {native_set=}')
+    def call_native(meth):
+        if native_get:
+            @functools.wraps(meth)
+            def _call_native(self, *args, **kwargs):
+                return getattr(self._native_widget, native_get)(*args, **kwargs)
+        if native_set:
+            @functools.wraps(meth)
+            def _call_native(self, *args, **kwargs):
+                getattr(self._native_widget, native_set)(*args, **kwargs)
+        return _call_native
+    return call_native
+
 class _AComponent:
@@ -207,2 +224,3 @@ def set_component_name(self, new_ctrl_name: str):
     @property
+    @_native(native_get='IsShown')
     def visible(self) -> bool:
@@ -212,5 +230,5 @@ def visible(self) -> bool:
         :return: True if this component is currently visible."""
-        return self._native_widget.IsShown()
 
     @visible.setter
+    @_native(native_set='Show')
     def visible(self, is_visible: bool):
@@ -219,3 +237,2 @@ def visible(self, is_visible: bool):
         :param is_visible: Whether or not to show this component."""
-        self._native_widget.Show(is_visible)
 
```
Has the advantage of pruning and singling out most boilerplate uses of `_native_widget`, can be used to control access in a central place and can be conditionally passed different functions for different backends (I am not familiar with any so this may not apply but I guess we should have a  _native_widgtet too and access similar methods - if not pass different functions). One default is harder typing - in particular we cant control click on the gui method and jump to the stub but this is still a couple clicks away (plus we should not be revisiting much the code in gui once frozen).

**Comment by lojack5** [comment]
I've been slowly working on a similar (more for a generic solution) attack on this same idea.  I like the approach.  My pet project is taking a lot of work because I'm really trying to get the IDE/static type-checkers to still be able to do some helpful work for me.  I want the typechecker to be able to point out when the method I'm forwarding to actually does have the same signature of the one I'm wrapping (just passing in the name of the method basically removes all ability of IDEs/etc to help you out there).  It's a hard problem getting *that* working properly, especially with all the different types of functions that could be accessed on an instance (bound method, unbound method, class method, static method, attribute that's a function but not an instance method, *properties*).  Anyway, it's a fun little problem to work on fleshing out all those details, who know's if I'll ever get it to a good working state though.

That aside, this is good enough for it's purposes :)

**Comment by Utumno** [comment]
An interesting typing exercise (passed when minimal to no code is added :P) would be to use typing for the events handling in our gui library - useful judging from the couple wrong params I passed to the on_drag callbacks recently.

**Comment by lojack5** [comment]
Not too hard to do, but would require a least one type definition when first defining the event handler.  It can be done in a way to even work with lambdas properly (which are notoriously hard to get to work with type information).  I'll mock something up, but they do end up using `TypeVar` and `TypeVarTuple`, which means we should hold off till 3.12 to make it much cleaner [PEP 695](https://docs.python.org/release/3.12.0/whatsnew/3.12.html#whatsnew312-pep695) is a godsend for cleaning up most of the boilerplate cruft involved.

**Comment by lojack5** [comment]
Ok, [lojack-typed-eventhandler](https://github.com/wrye-bash/wrye-bash/tree/lojack-typed-eventhandler) is an example of what can be done (requires python 3.12, hence the CI failure).  There's one thing blocking this really being useful, and one edge case that slightly hinders usability:

Blocker - wxPython type stubs don't have return type information.  Since most of our argument parsers are calling wx methods on event objects, the typechecker/IDE can't properly infer the types for the event handler.
Minor niggle - lambda functions defined inline (inside the call to `_AComponent._evt_handler` work well, because they can get type information from `_evt_handler` itself.  Lambda functions defined elsewhere have almost no useable type information, because the `event` parameter has no type information in that context.

I've already wanted to revisit the wxPython typestubs (they're autogenerated with a custom python script), this might make me go back and actually work on it.

**Comment by lojack5** [comment]
If you just want to play around on 3.12, everything seems to work find out of the box, we just can't package (pygit2 not available) and PDFs aren't shown (pymupdf not available).
